packages/server/src/status/service.ts

```diff
@@ -16,70 +17,102 @@ export class StatusService extends Service {
+      return status
     }
 
     return undefined
   }
 
-  async addError(conduitId: uuid, error: Error) {
+  async updateInitializedOn(conduitId: uuid, date: Date | undefined) {
     await this.distributed.using(`lock_dyn_status::${conduitId}`, async () => {
-      const numberOfErrors = await this.getNumberOfErrors(conduitId)
-      const formattedError = this.formatError(error)
+      if (!(await this.get(conduitId))) {
+        await this.create(conduitId)
+      }
 
-      if (numberOfErrors === undefined) {
-        await this.query().insert({ conduitId, numberOfErrors: 1, lastError: formattedError })
+      await this.query()
+        .update({ initializedOn: date ? this.db.setDate(date) : null })
+        .where({ conduitId })
+      this.cache.del(conduitId, true)
+    })
+  }
 
-        this.cache.set(conduitId, 1, undefined, true)
-      } else {
-        await this.query()
-          .update({ numberOfErrors: numberOfErrors + 1, lastError: formattedError })
-          .where({ conduitId })
+  async addError(conduitId: uuid, error: Error) {
+    await this.distributed.using(`lock_dyn_status::${conduitId}`, async () => {
+      const status = (await this.get(conduitId)) || (await this.create(conduitId))
+      const formattedError = this.formatError(error)
 
-        this.cache.set(conduitId, numberOfErrors + 1, undefined, true)
-      }
+      await this.query()
+        .update({ numberOfErrors: status.numberOfErrors + 1, lastError: formattedError })
+        .where({ conduitId })
+      this.cache.del(conduitId, true)
     })
   }
 
   async clearErrors(conduitId: uuid) {
     await this.distributed.using(`lock_dyn_status::${conduitId}`, async () => {
-      const numberOfErrors = await this.getNumberOfErrors(conduitId)
-
-      if (numberOfErrors && numberOfErrors > 0) {
-        await this.query().update({ numberOfErrors: 0, lastError: null }).where({ conduitId })
-
-        this.cache.del(conduitId, true)
+      const status = await this.get(conduitId)
+      if (!status) {
+        return
       }
-    })
-  }
 
-  private query() {
-    return this.db.knex(this.table.id)
+      await this.query().update({ numberOfErrors: 0, lastError: null }).where({ conduitId })
+      this.cache.del(conduitId, true)
+    })
   }
 
-  private async onConduitDeleted(conduitId: string) {
-    this.cache.del(conduitId, true)
+  async listOutdated(tolerance: number, maxAllowedFailures: number, limit: number): Promise<ConduitStatus[]> {
+    return (
+      (
+        await this.query()
+          // we exclude lastError because it migth make the query slow
+          .select('conduitId', 'numberOfErrors', 'initializedOn')
+          .where('numberOfErrors', '<', maxAllowedFailures)
+          .andWhere((q) =>
+            q
+              .where('initializedOn', '<', this.db.setDate(new Date(Date.now() - tolerance))!)
+              .orWhereNull('initializedOn')
+          )
+          .limit(limit)
+      ).map((x) => this.deserialize(x))
```

Comment:
What if the query returns no result?

Reply:
It that case the query returns an empty array, so the .map still works